analyzers/src/SonarAnalyzer.CSharp/Rules/ParameterTypeShouldMatchRouteTypeConstraint.cs

```diff
@@ -0,0 +1,238 @@
+                            property.ToAdditionalLocation(),
+                            property.ToAdditionalProperties(),
+                            property.ToPrimaryMessage()));
+                    }
+                },
+                SymbolKind.NamedType);
+        });
+
+    private static IReadOnlyList<PropertyTypeMismatch> GetPropertyTypeMismatches(INamedTypeSymbol classSymbol, Compilation compilation)
+    {
+        var routeParams = GetRouteParametersWithValidConstraint(classSymbol);
+
+        if (routeParams.Count == 0)
+        {
+            return Array.Empty<PropertyTypeMismatch>();
+        }
+
+        return classSymbol
+            .GetMembers()
+            .Where(IsPropertyWithParameterAttributeInRoute)
+            .SelectMany(property => routeParams[property.Name].Select(routeParam => new { RouteParam = routeParam, Property = property }))
+            .Where(x => !IsTypeMatchRouteConstraint(x.Property.GetSymbolType(), x.RouteParam.Constraint, compilation))
+            .SelectMany(x => x.Property.DeclaringSyntaxReferences
+                .Where(r => r.GetSyntax() is PropertyDeclarationSyntax)
+                .Select(r => new PropertyTypeMismatch(((PropertyDeclarationSyntax)r.GetSyntax()).Type, x.RouteParam.Constraint, x.RouteParam.FromRoute, x.RouteParam.RouteParamLocation)))
+            .ToList();
+
+        bool IsPropertyWithParameterAttributeInRoute(ISymbol member) =>
+            member.Kind is SymbolKind.Property && member.HasAttribute(KnownType.Microsoft_AspNetCore_Components_ParameterAttribute) && routeParams.ContainsKey(member.Name);
+    }
+
+    private static bool IsTypeMatchRouteConstraint(ITypeSymbol type, string routeConstraintType, Compilation compilation)
+    {
+        if (type.IsNullableValueType())
+        {
+            type = ((INamedTypeSymbol)type).TypeArguments[0];
+        }
+
+        return ConstraintMapping.ContainsKey(routeConstraintType) && ConstraintMapping[routeConstraintType].Matches(type, compilation);
+    }
+
+    private static Dictionary<string, List<RouteParameter>> GetRouteParametersWithValidConstraint(INamedTypeSymbol classDeclaration)
+    {
+        var routeParameters = new Dictionary<string, List<RouteParameter>>(StringComparer.InvariantCultureIgnoreCase);
+
+        var routeAttributeDataList = classDeclaration.GetAttributes()
+            .Where(x => KnownType.Microsoft_AspNetCore_Components_RouteAttribute.Matches(x.AttributeClass));
+
+        foreach (var routeAttributeData in routeAttributeDataList)
+        {
+            var routeNode = routeAttributeData.ApplicationSyntaxReference.GetSyntax() as AttributeSyntax;
+
+            if (!routeAttributeData.TryGetAttributeValue<string>("template", out var route) || routeNode is null)
+            {
+                continue;
+            }
+
+            var routeParams = route
+                .Split('/')
+                .Where(segment => segment.StartsWith("{") && segment.EndsWith("}"))
```

Comment:
`.Where(x=> x.StartsWith("{") && x.EndsWith("}"))`
Non-blocking: can be done in a following PR.